src/systems/pincer.rs

```diff
@@ -0,0 +1,96 @@
+use amethyst::{
+    core::Transform,
+    ecs::{Entities, Entity, Join, ReadStorage, System, WriteStorage},
+    renderer::{palette::Srgba, resources::Tint},
+};
+
+use crate::components::{Collider, Direction, Motion, Pincer, PincerAi, PincerAi::Attacking};
+
+/// Maximum distance over which a pincer will track its target.
+/// When the distance between the pincer and its target becomes greater than this constant,
+/// the pincer will lose interest and will revert back to patrolling.
+/// TODO: Arbitrarily picked, maybe later properly tune this variable.
+const MAX_TRACKING_DISTANCE: f32 = 10000.0;
+
+pub struct PincerAiSystem;
+
+/// Execute Pincer AI logic.
+///
+/// Attack logic is simple. Check the direction of the target and change the sign of
+/// the pincer's velocity to make it run towards the target.
+///
+/// There is one exception; the pincer won't change its direction while the target is jumping.
+/// This is to give the player a fair chance to jump over the pincer without the pincer's
+/// millisecond-reflexes screwing up the player's plans.
+///
+/// Finally, this system changes the pincer's color tint depending on whether it's aggro or not.
+impl<'s> System<'s> for PincerAiSystem {
+    type SystemData = (
+        Entities<'s>,
+        WriteStorage<'s, Pincer>,
+        WriteStorage<'s, Direction>,
+        WriteStorage<'s, Motion>,
+        WriteStorage<'s, Tint>,
+        ReadStorage<'s, Transform>,
+        ReadStorage<'s, Collider>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, mut pincers, mut directions, mut motions, mut tints, transforms, colliders): Self::SystemData,
+    ) {
+        let attacking_pincers = (&entities, &mut pincers, &transforms)
```

Comment:
Great! Nice and simple. And now the player has enough time to evade.

One change I would like to suggest though. Right now this system is performing the following 3 steps:
- Loop through all pincers and filter out the ones that are in Attacking state. Set the state to Patrolling if distance is greater than max distance.
- Loop through all pincers and change the direction of the ones who are in filtered in the first step.
- Loop through all pincers and change the color for those who are in Attacking state.

I guess this can be done in a single loop. That is, loop through all pincers and:
- change color for those who are in Attacking state.
- change direction only if the target is on ground.
- if distance between target and pincer is greater than the max distance then set pincer's state to Patrolling.

This way we don't have to loop 3 times. Let me know if you see any potential issues with the suggested approach.

Reply:
Yes, currently it should be possible to do it all in one loop. That would probably look a bit cleaner.

The reason it ended up like this is because I initially thought I needed the marine's Motion component to be able to see whether it was jumping or not. (Motion has a 'has_jumped' field) Since I was already mutably borrowing the Motion storage for the pincer, I couldn't also borrow it for the marine. Doing it in multiple loops was the only way to do it that I could see.

However, it turns out I didn't need Motion but Collider. Perhaps in the future we'll need to borrow more of the marine's components; that's why I left it like this. 

It's up to you, I can make it into one loop or we can leave it like this.

Reply:
I don't have a strong preference either way.

Reply:
Makes sense. Though I would prefer one loop for now. We can break it down later if required.

Reply:
Okay, done.